enhancements/storage/vsphere-driver-configuration.md

```diff
@@ -0,0 +1,308 @@
+follow API conventions for adding new fields to an existing CRD as described here: https://github.com/openshift/enhancements/blob/2894d936eca7c1fcc9cf38e7dc973bbdfa1d88ff/dev-guide/api-conventions.md#new-field-added-to-an-existing-crd
+
+#### Option 1: Add a specific API field for the maximum limit of snapshots
+
+This solution would require adding a new field in the ClusterCSIDriver CRD that would hold the maximum limit of snapshots
+and follow the usual API review process. A considerable downside is that there could be other related configuration
+options for a CSI driver - for example in vSphere, the maximum number of volumes that can be created per volume can be
+configured in 3 different ways: `global-max-snapshots-per-block-volume`, `granular-max-snapshots-per-block-volume-vsan`,
+`granular-max-snapshots-per-block-volume-vvol`. In this case we would pick only the first one which is a global setting and
+overrides the other two.
+
+Example:
+
+1. Cluster administrator sets integer value of 10 in `clustercsidriver.spec.driverConfig.vSphereConfig.snapshotLimit`
+2. The driver operator reconciles the ClusterCSIDriver object and propagates the option to a valid destination. This can 
+be a config ini file of the driver that will be updated to contain `global-max-snapshots-per-block-volume = 10` option
+under `[Snapshot]` stanza.
+3. The driver controller pods are restarted and new configuration loaded.
+4. Users can now create VolumeSnapshot 10 times per volume, limit for VVOL and VSAN is also 10 and can not be changed by 
+the administrator.
+
+- Pros:
+  - Simple and straightforward
+  - Very stable and easy to maintain
+- Cons:
+  - Only supports a single configuration option
+  - Does not scale with number of RFE requests due API review process
+  - Is limiting for users because they can't use any feature which we don't explicitly allow
+
+#### Option 2: Configuration via generic map in a dedicated field of ClusterCSIDriver
+
+This solution would require having a new field in the ClusterCSIDriver CRD that would hold the all configuration options
+as a map (`map[string]string`), something like `clustercsidriver.spec.vSphere.configuration`. Then we would merge it into
+any other valid destination for the driver - command line arguments, environment variables, ini file, driver
+feature gate ConfigMap, etc.
+
+Alternatively the configuration could be provided as a ConfigMap directly and the driver operator would reconcile it and
+ensure the options are propagated to valid destinations.
+
+For this approach it would be reasonable to have some kind of validation if unsupported options were used and warn users
+if it happens. This could be in console, alerts, logs or must-gather.
+
+Example:
+
+1. Cluster administrator creates a ClusterCSIDriver object with the configuration map set to `{"maxSnapshots": "10"}`
+under a generic `map[string]string` field (`clustercsidriver.spec.vSphere.configuration`)
```

Comment:
Here is what I am thinking:

We break down the singular `configuration` field into two:

`clusterCSIDriver.Spec.vSphere.configParameters` and `clusterCSIDriver.Spec.vSphere.featureFlags` . These are still a `map[string]string` - but internally in vSphere operator we keep track of configParameters/featureFlags we support (like a whitelist) and only whitelisted fields can be applied and have an effect. We can ignore unknown config/featureFlags and log a message or something.

This is still very similar to option#1 but only benefit it provides is, say after a rebase we want to support a new config option, then we don't necessarily have to bump `openshift/api`. We just update the whitelist inside vmware-vsphere-csi-driver. 

We are not in a position to support any random config parameter, because we don't even know how that will be applied, so we can only support a sanctioned list of configs and that is where whitelist might be helpful.

Reply:
Having a separate field for `featureFlags` sounds good, I'm adding it.

Whitelist is a type of validation, at first I was not too specific about this - adding it.

The benefit of this approach (not having to bump `openshift/api`) is already mentioned under `Pros:`.